content/blog/2019-09-25-type_checker.md

```diff
@@ -0,0 +1,88 @@
+
+The error raised by the type checker outputs the line at which we find the first operation breaking the type check rule and mentions the type of error found.
+
+
+### Implementation
+
+We have a first pass in the algorithm which collects a list of labels and ensures that each label is unique. If there are multiple labels with the same name, we throw an error at this point. This also helps us create a set of label names which is used to check for valid strings (of labels) during the second pass for control flow operations.  In the second pass, we go over each instruction and check for various type errors. Some fundamental checks in this process are:
+
+1. Invalid instructions- This ensures that all the arguments and destination (if applicable) of an instruction are available, which is given by the length of instruction.
+2. Argument and destination type- For various operations we check if arguments and destination variables have the correct type. Something like `bool d = a + b` where `a,b` are integers would raise an error for the destination variable.
+3. Redefined variables- We check if the destination variable has already been assigned to a different type in that context. We do allow const instructions on the same type as it can be treated as an assignment operation rather than a definition. So `int a =2; bool a = true` is not allowed but `int a =2; int a =5` is allowed. We do this by keeping a set of variables of each type (int and bool) defined in the function. This helps while checking existing definitions and possible redefinition errors.
+4. Undefined variables- We check if the arguments to the instruction have defined variables using the set of variables mentioned before. 
+
+
+
+## Hardest parts during the implementation
+
+The type checker implementation, though straightforward, had a few challenges.
+
+1. The typing rules need to be designed carefully for each operation, ranging from arithmetic, boolean to control instructions. Take the branch instruction as an example, say we have "br b left right". The type checker needs to go through all the existing boolean variables to ensure that b is predefined and left and right are valid labels in the code snippet. Also, though not currently supported in Bril, the instructions could be nested. Thus, it is important to maintain modularity of arithmetic/boolean and control flow checking so that future updates could be made easily to support recursive checking.
```

Comment:
> Take the branch instruction as an example, say we have "br b left right".

This is a comma splice—consider a colon instead of a comma. And use Markdown code backticks instead of quotes.